### DISCUSSION TITLE
question: how promtail code and agent are synchronized

### DISCUSSION THREAD

**Discussion by eraac** [grafana-cold-storage/agent#3842] [Q&A] [answered]
Hello,

From my understand, the Grafana Agent include the Promtail code to perform feature related to Loki/Logs, but I can't figure out how agent can include new Promtail features?

I've push a MR to Loki https://github.com/grafana/loki/pull/9044 but now I'm not sure agent got the new feature as well along with a new release?

- Does the change must be (also) apply in this repository? (which I can do too)
- Does there is a magic "sync script"?

Thanks

**Comment by mattdurham** [comment, marked as answer]
There are currently two paths to getting promtail code into the Agent.

The first is for static mode in which we import `github.com/grafana/loki`, if you look at our [go.mod](https://github.com/grafana/agent/blob/main/go.mod) file you will see a line `github.com/grafana/loki v1.6.2-0.20230414223651-220cbdd4f172 // k146 branch` this specifies we have pulled in the k146 branch. Normally we update this every few months or whenever requests for features start to accumulate.

The second way for Flow mode is we are lift and shifting the promtail code directly into the Agent. If you look at [component/loki/source/docker/internal/dockertarget/target.go](https://github.com/grafana/agent/blob/main/component/loki/source/docker/internal/dockertarget/target.go) you will notice a disclaimer where we are copying the code with light edits.

If you wanted to create a PR to update to k149 we would definitely take a look at that.
